zwave_js_server/client.py

```diff
@@ -319,6 +339,81 @@ def end_recording_messages(self) -> list[dict]:
 
         return list(data)
 
+    @property
+    def server_logging_enabled(self) -> bool:
+        """Return whether server logging is currently enabled."""
+        return self._server_logging_enabled
+
+    async def enable_server_logging(self) -> None:
+        """
+        Enable logging from the server.
+
+        Embeds server logs in the library's logs.
+        """
+        if not self.connected or not self.driver:
+            raise InvalidState(
+                "Can't enable server logging when not connected to server"
+            )
+        if (log_level := self.driver.log_config.level) and (
+            level := LOG_LEVEL_MAP[log_level]
+        ) < self._logger.level:
+            self._logger.info(
+                (
+                    "Server logging is currently more verbose than library logging so "
+                    "setting library log level to match."
+                )
+            )
+            self._logger.setLevel(level)
+
+        self._server_logging_enabled = True
+
+        def handle_server_logs(event: dict) -> None:
+            """Handle driver `logging` event."""
+            log_msg: LogMessage = event["log_message"]
+            log_level = LogLevel(log_msg.level)
+            logging.getLogger(f"{__package__}.server").log(
+                LOG_LEVEL_MAP[log_level],
+                "%s:\n%s",
+                log_msg.timestamp,
+                "\n".join(log_msg.formatted_message),
+            )
+
+        def handle_log_config_updates(event: dict) -> None:
+            """Handle driver `log config updated` events."""
+            if (log_level := event["level"]) and (
+                level := LOG_LEVEL_MAP[log_level]
+            ) < self._logger.level:
+                self._logger.info(
+                    (
+                        "Server logging is currently more verbose than library "
+                        "logging so setting library log level to match."
+                    )
+                )
+                self._logger.setLevel(level)
+
+        self._server_logger_unsubs = [
+            self.driver.on("logging", handle_server_logs),
+            self.driver.on("log config updated", handle_log_config_updates),
+        ]
+
+        await self.async_start_listening_logs()
+
+    async def disable_server_logging(self) -> None:
+        """Disable logging from the server."""
+        if not self.connected or not self.driver:
+            raise InvalidState(
+                "Can't disable server logging when not connected to server"
+            )
+        if not self._server_logging_enabled:
+            raise InvalidState("Server logging is already disabled")
+        if not self._server_logger_unsubs:
+            raise InvalidState("There is no logging event listener to unsubscribe")
+
+        for unsub in self._server_logger_unsubs.copy():
```

Comment:
copied the Event listener logic without thinking but it's unnecessary here because the subscriptions are managed internally and will never get modified outside of these